runtime: Simplify slice growing/appending code

lpereira opened the pull request:
Refactor the slice appending function to rely on the slice growing function, and remove branches/loops to use a branchfree variant.

CC @dgryski

lpereira committed:
runtime: Simplify slice growing/appending code

Refactor the slice appending function to rely on the slice growing
function, and remove branches/loops to use a branchfree variant.

Signed-off-by: L. Pereira <l.pereira@fastly.com>

dgryski commented:
Some quick testing shows this makes the multiplication factor 1.125.  However the expression doesn't seem to work for `newCap <= 8` so I think we need to make sure the edge cases are handled properly.

dgryski commented:
I'm not sure this actually saves us anything.  I threw together a quick simulation and it seems that the new algorithm requires more calls to "grow" (since we're not growing as much, we need to do it more frequently, requiring more CPU) *and* ends up generating more garbage, because the additional calls to grow means we have to throw more stuff away.

You can play with my sample code here: https://go.dev/play/p/QcHiMqnGlym

lpereira commented:
Yup, that seems to be the case from some of the testing I did last week to understand why some tests were failing. I'm going to send a new version that does pretty much the same thing as before but keeps the cleanup.

On Mon, Jul 22, 2024, at 11:53 AM, Damian Gryski wrote:
> 
> 
> I'm not sure this actually saves us anything. I threw together a quick simulation and it seems that the new algorithm requires more calls to "grow" (since we're not growing as much, we need to do it more frequently, requiring more CPU) *and* ends up generating more garbage, because the additional calls to grow means we have to throw more stuff away.
> 
> You can play with my sample code here: https://go.dev/play/p/QcHiMqnGlym
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/tinygo-org/tinygo/pull/4287#issuecomment-2243609999>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAADVGKW2UBXOAKSZRH5APLZNVIKBAVCNFSM6AAAAABI7FV6YCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENBTGYYDSOJZHE>.
> You are receiving this because you authored the thread.Message ID: ***@***.***>
>

lpereira commented:
Alright, @dgryski, pushed a new version, rebased on the current `dev` branch, that goes back to the "align to the next power of two" code, but using a branchfree version instead.

lpereira commented:
The failure inside the `reflect` test is very weird.

lpereira committed:
runtime: Remove one branch in sliceAppend()

Both branches were equivalent, so guard the overall logic in
sliceAppend() with the more general condition.

Signed-off-by: L. Pereira <l.pereira@fastly.com>

lpereira committed:
runtime: Simplify slice growing calculation

Use `bits.Len()` rather than `32 - bits.LeadingZeros32()`.  They're
equivalent, but the Len version is a bit easier to read.

Signed-off-by: L. Pereira <l.pereira@fastly.com>

lpereira committed:
reflect: Always call sliceGrow() in extendSlice()

sliceGrow() will return the old slice if its capacity is large enough.

Signed-off-by: L. Pereira <l.pereira@fastly.com>

cee-dub commented:
The Index field of "want" for the reflect field `F` looks incorrect, it is field 1, but want is `0,0`.

ldemailly commented on src/runtime/slice.go:
```diff
@@ -54,29 +36,21 @@ func sliceCopy(dst, src unsafe.Pointer, dstLen, srcLen uintptr, elemSize uintptr
-	for oldCap < newCap {
-		oldCap *= 2
-	}
+	// This can be made more memory-efficient by multiplying by some other constant, such as 1.5,
```
It's 2x only for small slices (and then toward 1.25x for big ones) in big go - should it also be same for tinygo or expectation is always "small"-ish in tinygo ?

https://github.com/golang/go/blob/e76353d5a923dbc5e22713267104b56a2c856302/src/runtime/slice.go#L304

lpereira replied:
Good idea!  Doing 1.5x for slices over 64 elements now.

ldemailly commented on src/runtime/slice.go:
```diff
@@ -54,29 +36,21 @@ func sliceCopy(dst, src unsafe.Pointer, dstLen, srcLen uintptr, elemSize uintptr
+	// This can be made more memory-efficient by multiplying by some other constant, such as 1.5,
+	// which seems to be allowed by the Go language specification (but this can be observed by
+	// programs).
+	newCap = 1 << bits.Len32(uint32(newCap))
```
so tinygo is 32bits only? 

also tinygo doesn't need to account for malloc headers like big go?

lpereira replied:
Changed this to use `bits.Len()`, passing an `uint` instead of `uint32`.  Hopefully this should work on 64-bit too.

ldemailly replied:
Thanks,

and for the other part (which may not be applicable to tiny go, depends on what kind of heap allocator there is but I think in big go they subtract some meta info size so for instance (making actual numbers up) when requesting 3500 bytes you get a full 4096-headers page of memory instead of 4096+fragment)

ps: I have no idea of tinygo's internal

aykevl replied:
Whether there is metadata or not depends on `-gc=precise` vs `-gc=conservative`. But that should be an implementation detail of the GC/heap: all the slice code needs to know is that there's the `alloc` function and it's not allowed to write outside the returned memory area.

ldemailly commented on src/runtime/slice.go:
```diff
@@ -54,29 +36,21 @@ func sliceCopy(dst, src unsafe.Pointer, dstLen, srcLen uintptr, elemSize uintptr
+	newCap = 1 << bits.Len32(uint32(newCap))
 
-	buf := alloc(oldCap*elemSize, nil)
+	buf := alloc(newCap*elemSize, nil)
```
should this have some overflow check?

lpereira replied:
Yes, it should. Even calculating the new `newCap` can overflow.  However, handling overflow conditions doesn't seem to be something that's done pervasively in TinyGo, so I didn't do it here so this is closer to the rest of the code.

aykevl replied:
I think it is very unlikely to overflow in practice, but yes it would be a good idea to have some sort of check for this.
For 32-bit and 64-bit systems it would be possible to use math/bits for this: https://pkg.go.dev/math/bits#Add32
But maybe there are universal expressions for detecting such an overflow condition efficiently.

ldemailly reviewed (commented):
lgtm. nice to share the code in grow. just some (non merge blocking) questions

lpereira commented:
> The Index field of "want" for the reflect field `F` looks incorrect, it is field 1, but want is `0,0`.

Right, and I have no idea why!  I've spent some time in the reflect code, and @dgryski too, but we don't yet know what's happening here.  Our best guesses is that there's either memory corruption somewhere, or that, although the code seems correct, its behavior is different from before, and the reflect code is probably depending on that somehow.

cee-dub commented:
Hmm, I was suggesting/agreeing with @dgryski that the test may be faulty and the old implementation matched. I think the expectation looks incorrect, though I have very little experience or deep knowledge of this code.

lpereira commented:
> Hmm, I was suggesting/agreeing with @dgryski that the test may be faulty and the old implementation matched. I think the expectation looks incorrect, though I have very little experience or deep knowledge of this code.

If I understood the test correctly, it seems correct. The index is a slice of indices, starting with the current type, and following any nested type. So an index of `[]int{0}` means the field is the first in that struct; and an index of `[]int{0, 1}` means it's the second of the first of that struct.

The test defines two structs: `Rec1` and `Rec2`; `Rec1` has a pointer to `Rec2`, and `Rec2` has a `F string` and a pointer to `Rec1`. To find `F` from `Rec1`, you need to go through the first element, and then find the first element (`[]int{0, 0}`), which seems to agree with the test.

I might be mistaken about how this works, though.

cee-dub commented:
Reading more of #tinygo-dev, feel free to ignore me. I didn't look closely enough at the `Index` type in the test.

lpereira commented:
So I copied the failing test to its own file, and `reflect.VisibleFields()` gives the right output, agreeing with Big Go.  So the problem seems to be in the actual test driver somewhere.

lpereira commented:
Fixed the `reflect` test; let's see if CI likes it.

lpereira commented:
Alright, `archive/zip` failing is probably the `reflect.DeepEqual()` calls inside the tests, most likely some unbounded recursion or something of the sort.

dgryski commented:
@lpereira Ok, so it does seem like the bug is somewhere in `reflect` then.  I'm going to try to hunt down the bug looking at that function.

dgryski commented:
This fixes the `reflect` test:
```
diff --git a/src/reflect/type.go b/src/reflect/type.go
index 1356f67c..32bf68d7 100644
--- a/src/reflect/type.go
+++ b/src/reflect/type.go
@@ -774,7 +774,7 @@ func (t *rawType) rawFieldByNameFunc(match func(string) bool) (rawStructField, [
                                if match(name) {
                                        found = append(found, result{
                                                rawStructFieldFromPointer(descriptor, field.fieldType, data, flagsByte, name, offset),
-                                               append(ll.index, int(i)),
+                                               append(ll.index[:len(ll.index):len(ll.index)], int(i)),
                                        })
                                }

@@ -787,7 +787,7 @@ func (t *rawType) rawFieldByNameFunc(match func(string) bool) (rawStructField, [

                                        nextlevel = append(nextlevel, fieldWalker{
                                                t:     embedded,
-                                               index: append(ll.index, int(i)),
+                                               index: append(ll.index[:len(ll.index):len(ll.index)], int(i)),
                                        })
                                }

```

I'll get this in as another PR and you can rebase against that with only the slice allocation changes.

dgryski committed:
reflect: rawFieldByNameFunc: copy index slice to avoid later overwrites

dgryski commented:
https://github.com/tinygo-org/tinygo/pull/4367

lpereira commented:
I rebased this on top of #4387 (please merge it first!), and removed the commit changing `reflect`.  CI is still failing with some error I was unable to track down.

dgryski commented:
I'll look at the `archive/zip` failure this evening.

dgryski commented:
The `archive/zip` failure seems to be related to the stepped doubling logic function.

dgryski commented:
Nope; failure seems intermittent.

dgryski commented:
This is due to memory fragmentation which (annoyingly) has different behaviour with the new growth algorithm.  A smarter allocator that has free lists for small allocation sizes to reduce fragmentation would help here, but that's outside the scope of this PR.

lpereira commented:
> A smarter allocator that has free lists for small allocation sizes to reduce fragmentation would help here, but that's outside the scope of this PR.

Agreed.  I've rebased on the current `dev` branch and changed the code to have only one overallocation strategy until we have a better allocator.  This won't reduce the amount of memory used by slices, but it simplifies the code quite a bit, which is already a gain IMHO.

ldemailly commented:
fragmentation because of not taking headers/meta into account was my point in https://github.com/tinygo-org/tinygo/pull/4287#discussion_r1692116535

lpereira commented:
sizediff failure is unrelated.

dgryski reviewed (approved):
LGTM

dgryski commented:
Merging.  Will close my PR that has the reflect fix.

aykevl commented on src/reflect/type.go:
```diff
@@ -774,7 +774,7 @@ func (t *rawType) rawFieldByNameFunc(match func(string) bool) (rawStructField, [
 					found = append(found, result{
 						rawStructFieldFromPointer(descriptor, field.fieldType, data, flagsByte, name, offset),
-						append(ll.index, int(i)),
+						append(ll.index[:len(ll.index):len(ll.index)], int(i)),
```
This will force a new allocation on every `append`, is that intentional?

dgryski replied:
Yes.  This fixes a test failure in `reflect` that was triggered by the new slice allocation code which made this sometimes be a copy and sometimes not.  (The old growth code made this always a copy in the test, but the new growth code expanded it more and so sometimes reused the old slice, which then had entries in the slice overwritten by later loop iterations, causing the faiure.)
